include/swift/AST/DiagnosticsAll.def

```diff
@@ -42,6 +42,21 @@
 #  define FIXIT(ID, Text, Signature)
 #endif
 
+#ifndef PUBLICERROR
+#  define PUBLICERROR(PUBLICID, ID, OPTIONS, TEXT, SIGNATURE) \
+  ERROR(ID, OPTIONS, TEXT, SIGNATURE)
```

Comment:
I would also suggest changing all diagnostics, or maybe all non-note diagnostics, to require a public ID, but allow "none" as a valid choice. But maybe this way makes more sense, since it forces people to think before making something public.

I do think the compiler ID should go first, not the public ID.

Reply:
I thought about this a bit more and think I've come up with a pretty nice solution here. I introduced two new macros, INTERNALERROR and INTERNALWARNING. These expand to ERROR and WARNING respectively, but they help document the fact that somebody has audited this diag and decided it shouldn't be public. Now the options are:

- ERROR: not public, hasn't been audited, implicitly has a public id of 'none'
- INTERNALERROR: not public, has been audited, implicitly has a public id of 'none'
- PUBLICERROR: public, has been audited

This should hopefully make the incremental rollout less error-prone.

I agree the compiler id should come before the public one, that's been fixed.

As a sidenote, I also moved the macro definitions for the diagnostic files into their own header that can be included as needed. With this patch they'd become too large to maintain 10 identical copies, and there were already a few discrepancies.